block/produce.go

```diff
@@ -95,40 +96,47 @@ func (m *Manager) ProduceBlockLoop(ctx context.Context, bytesProducedC chan int)
 	}
 }
 
-func (m *Manager) ProduceAndGossipBlock(ctx context.Context, allowEmpty bool) (*types.Block, *types.Commit, error) {
-	block, commit, err := m.produceBlock(allowEmpty)
+func (m *Manager) ProduceApplyGossipLastBlock(ctx context.Context, nextProposerHash [32]byte) (block *types.Block, commit *types.Commit, err error) {
+	// check if block with nextProposerHash already produced and applied
+	h := m.State.Height()
+	block, err = m.Store.LoadBlock(h)
 	if err != nil {
-		return nil, nil, fmt.Errorf("produce block: %w", err)
+		return nil, nil, fmt.Errorf("load block: height: %d: %w", h, err)
 	}
-
-	if err := m.gossipBlock(ctx, *block, *commit); err != nil {
-		return nil, nil, fmt.Errorf("gossip block: %w", err)
+	if bytes.Equal(block.Header.NextSequencersHash[:], nextProposerHash[:]) {
+		m.logger.Debug("Last block already produced and applied")
```

Comment:
If it's already produced and applied, why is this function being called? I mean, when can this happen?

Reply:
last block produced but state update not sent yet